Implemented sinc function

leosartaj opened the pull request:
Implemented sinc class

Examples

```
>>> from sympy import sinc, oo, jn, Product, Symbol
>>> from sympy.abc import x
>>> sinc(x)
sinc(x)
```

Automated Evaluation

```
>>> sinc(0)
1
>>> sinc(oo)
0
```

Differentiation

```
>>> sinc(x).diff()
(x*cos(x) - sin(x))/x**2
```

Series Expansion

```
>>> sinc(x).series()
1 - x**2/6 + x**4/120 + O(x**6)
```

As zero'th order spherical Bessel Function

```
>>> sinc(x).rewrite(jn)
jn(0, x)
```

Product form

```
>>> k = Symbol('k', integer=True, positive=True)
>>> sinc(x).rewrite(Product)
Product(2**(-k)*cos(x), (k, 1, oo))
```

fixes #5051

@certik @asmeurer

aktech commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1568,6 +1568,97 @@ def taylor_term(n, x, *previous_terms):
+                return S.NaN
+
+        if arg is S.ComplexInfinity:
+            return S.ComplexInfinity
```
`sinc(zoo)` is probably indeterminate.

jksuom replied:
Yes, ``sinc`` has no limit at ``zoo`` ([an essential singularity](http://en.wikipedia.org/wiki/Essential_singularity)). I believe the value should be ``nan`` in SymPy.

aktech commented:
You could also add sine rewrite:
```python
def _eval_rewrite_as_sin(self, arg):
        return sin(arg)/arg
```

Otherwise, it looks good to me.

aktech commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1568,6 +1568,97 @@ def taylor_term(n, x, *previous_terms):
+    >>> sinc(x)
+    sinc(x)
+
+    Automated Evaluation
```
Automated Evaluation  --->  * Automated Evaluation 
(It would look better in docs)

aktech commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1568,6 +1568,97 @@ def taylor_term(n, x, *previous_terms):
+    >>> sinc(oo)
+    0
+
+    Differentiation
```
same

leosartaj commented:
I have made the changes @aktech @jksuom

jksuom commented:
I am worried about this
```
In [3]: s = Symbol('s', zero=True)

In [6]: sin(s*pi)
Out[6]: 0

In [7]: sinc(s*pi)
Out[7]: 0
```

jksuom commented:
References to ``Product(cos(arg)/2**k, (k, 1, S.Infinity))`` and ``Integral(sinc(t), (t, 0, z))`` would be useful.

toolforger commented:
Am 09.04.2015 um 11:48 schrieb Kalevi Suominen:
> I am worried about this
> ```
> In [3]: s = Symbol('s', zero=True)
>
> In [6]: sin(s*pi)
> Out[6]: 0
>
> In [7]: sinc(s*pi)
> Out[7]: 0
> ```

Right - sinc's singularity at zero is removable, but the value should be 
`1` (or `NaN` if SymPy does not remove removable singularities, dunno 
SymPy's policy for that).

leosartaj commented:
This should return 1

```
In [3]: s = Symbol('s', zero=True)

In [6]: sin(s*pi)
Out[6]: 0

In [7]: sinc(s*pi)
Out[7]: 1
```

I will fix this.

leosartaj commented:
I have fixed the following

```
>>> n = Symbol('n', integer=True)
>>> sinc(n*pi)
sinc(n*pi)
>>> s = Symbol('s', zero=True)
>>> sinc(s*pi)
1
>>> k = Symbol('k', integer=True, nonzero=True)
>>> sinc(k*pi)
0
```

@jksuom

jksuom commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1568,6 +1568,100 @@ def taylor_term(n, x, *previous_terms):
+        pi_coeff = _pi_coeff(arg)
+        if pi_coeff is not None:
+            if pi_coeff.is_integer:
+                if arg.is_nonzero:
```
This is probably not necessary any more as ``arg.is_zero`` is tested above.

leosartaj replied:
This is required for the case

```
>>> n = Symbol('n', integer=True) # may or may not be zero
>>> sinc(n*pi)
sinc(n*pi)
```

jksuom replied:
Yes, that is right. Though ``pi_coeff.is_nonzero`` might be more suitable as the assumption mechanism would not need to check ``pi.is_nonzero``. 
(Perhaps a comment could be added here.)

leosartaj replied:
From the docstring of _pi_coeff, it returns in the range [0, 2].

```
>>> from sympy.functions.elementary.trigonometric import _pi_coeff
>>> _pi_coeff(10*pi)
0
```

This will cause problems in the cases, like
```
>>> sinc(10*pi)
sinc(10*pi)
```

I think checking arg is better here.

jksuom replied:
Ok. I forgot it was normalized. That is not so useful here as in the case of ``sin()``.

jksuom commented:
I think that references to the formulas with Product and Integral would be really needed.

leosartaj commented:
I already did add them

```
>>> sinc(x).rewrite(Product)
Product(2**(-k)*cos(x), (k, 1, oo))
>>> Si(x).rewrite(sinc)
Integral(sinc(t), (t, 0, x))
```

jksuom commented:
I am sorry, perhaps I should have expressed myself more clearly. I meant references to the literature justifying these formulas.

leosartaj commented:
Fine, I will add some references. Also, will it be appropriate to have the following functionality

```
>>> sinc(x).rewrite(Si)
Derivative(Si, x)
```

jksuom commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1568,6 +1568,100 @@ def taylor_term(n, x, *previous_terms):
+                if arg.is_nonzero:
+                    return S.Zero
+            elif (2*pi_coeff).is_integer:
+                return S.NegativeOne**(pi_coeff - S.Half) / arg
```
Should there be a test exercising this line?

leosartaj replied:
I have added some tests, testing this piece of code.

jksuom commented:
> Also, will it be appropriate to have the following functionality
>>> sinc(x).rewrite(Si)
Derivative(Si, x)

I doubt if a derivative would qualify as a rewrite. My knowledge of SymPy policy is insufficient.
@asmeurer ?

leosartaj commented:
I have improved the docs. Have a look @jksuom.

jksuom commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -1568,6 +1568,101 @@ def taylor_term(n, x, *previous_terms):
+    def _eval_rewrite_as_Product(self, arg):
+        from sympy import Product
+        k = Symbol('k', Dummy=True, integer=True, positive=True)
+        return Product(cos(arg)/2**k, (k, 1, S.Infinity))
```
I have had doubts about whether this product formula should be called a rewrite as it is an infinite product. There seems to be no comparable rewrite method in SymPy. Besides, there are other infinite product formulas for ``sinc`` (and ``sin``). Are there some specific reasons for including this one?

leosartaj replied:
No nothing specific, I found it interesting. If not as a rewrite, then what is the other way to handle it?

jksuom replied:
Perhaps the formula should not belong to this PR at all. There are product formulas for many other functions, but they have not been included in SymPy. They probably do not fit well into its design.

leosartaj commented:
I have removed the Product rewrite.

leosartaj commented:
Is this good to go? (after squashing)

jksuom commented:
I think this can be squashed. It looks ok.

leosartaj committed:
Implemented sinc function

Examples
========

>>> from sympy import sinc, oo, jn, Product, Symbol
>>> from sympy.abc import x
>>> sinc(x)
sinc(x)

* Automated Evaluation

>>> sinc(0)
1
>>> sinc(oo)
0

* Differentiation

>>> sinc(x).diff()
(x*cos(x) - sin(x))/x**2

* Series Expansion

>>> sinc(x).series()
1 - x**2/6 + x**4/120 + O(x**6)

* As zero'th order spherical Bessel Function

>>> sinc(x).rewrite(jn)
jn(0, x)

References
==========

.. [1] http://en.wikipedia.org/wiki/Sinc_function

leosartaj commented:
I have squashed the commits @jksuom.
